AaC Style Guide (#613)
jondavid-black/AaC: docs/docs/user_guide/style_guide.md

The following definitions follow the style of naming fields and attributes in camelCase.

```yaml
enum:
    name: ExampleSubStructure
    fields:
        - name: msgContent
          type: string
---
schema:
    name: ExampleInterfaceMessage
    fields:
        - name: msgMetadata
          type: string
        - name: msgData
          type: ExampleInterfaceMessage
```

Users leveraging the `ExampleInterfaceMessage` will have a data structure with fields such as `ExampleInterfaceMessage.msgMetaData` and `ExampleInterfaceMessage.msgData.msgContent`. Alternatively, if a user didn't follow the style for fields and attributes, our example definitions might look like:

```yaml
enum:
    name: ExampleSubStructure
    fields:
        - name: msg_content
          type: string
---
schema:
    name: ExampleInterfaceMessage
    fields:
        - name: msg_Metadata
          type: string
        - name: MessageData
          type: ExampleInterfaceMessage
```

> The resulting data structure `ExampleInterfaceMessage` will now have an inconsistent naming convention in its structure with fields such as `ExampleInterfaceMessage.msg_Metadata` and `ExampleInterfaceMessage.MessageData.msg_content`. The inconsistent naming of fields can easily spiral into data structures and models that are hard to navigate or interact with for users and plugin developers.

Reviewer: My example was auto-complete but, generally, I just meant a written statement talking about preemptively addressing the fact that tooling can alleviate this but why it's still ideal to conform to a consistent format.

    fields:
        - name: enumField
          type: sample_enum
```

In this example, a user unfamiliar with the project can easily mistake `simple_enum` as a primitive type.

## Enumeration Values
Enumeration values in AaC are all UPPSERCASE, following the common style for enumeration values and constants in programming languages such as Python and Java.

```yaml
enum:
    name: SampleEnums
    values:
     - VALUE_1
     - VALUE_2
     - VALUE_3
```

## Primitive Types
Primitive types in AaC are defined by the enumeration values in the `Primitives` definition. Each enum value respresents a uniquely special set of enumeration values that define the core AaC language's primitive types. Users who extend the primitive values should maintain the same style of snake_case and keep primitive type names succint names that are descriptive but short.

## Root Keys
Like the Primitive Types, the available root keys in AaC are defined in the definition `Root`. Each field entry in the `Root` definition specifies the root key (the field's name) and a corresponding schema definition (the field type). Additional root keys can be defined by users via an extension to the `roots` definition.

> Like primitive types, root keys should follow the snake_case style, but emphasize succint, descriptive names. Example root keys are `import`, `schema`, `model`, etc. If a one-word type isn't descriptive enough for your root key, keeping it succint (e.g. `spring_boot_service`, `cloud_gateway`) is preferable. Long-named root keys, or roots keys that aren't snake_case, will not conform to the standard employed by the Core Specification and first-party plugin making it less obvious to users what the key is.

Reviewer: ```suggestion
Like primitive types, root keys should follow the `snake_case` style, but emphasize succinct, descriptive names. Example root keys are `import`, `schema`, `model`, etc. If a one-word type isn't descriptive enough for your root key, keeping it as succinct as possible (e.g. `spring_boot_service`, `cloud_gateway`) is preferable. Long-named root keys, or roots keys that aren't `snake_case`, will not conform to the standard employed by the Core Specification and first-party plugins and could introduce confusion about what the root key is.
```

# Organizing Imports
Each AaC file has a file-wide import declaration definition. Using an `import` definition allows users to reference definitions provided via other user files and can be used to decompose large, complex models into multiple files.

Because `import` decalarations are file-wide, these clauses should be the first clause in AaC files. There is no internal logic that dictates the order of the entries in an AaC file, but maintaining the `import` clause as the first definition in AaC files makes the file-wide dependencies immediately visible and informs readers on the dependencies of file.

An example file with an import clause at the top:

```yaml

> import:
>     - ../interface/service_a_messages.aac

Reviewer: ```suggestion
import:
    files:
        - ../interface/service_a_messages.aac
```

The `import` keyword supports a list of relative or absolute paths to other user-defined files in the overall project. Users are not confined to a working directory or a single-flat directory for their AaC files. The files can be distributed in directories and subdirectories, or even alongside their related components in existing software projects.

> It's recommended that your AaC file structure avoids circular imports (e.g. File A imports File B, File B imports File C, and File C imports File A). AaC has some internal checks to de-conflict circular imports, but circular imports may cause unexpected behavior and indicate poor project  organization.

Reviewer: ```suggestion
It's recommended that your AaC file structure avoids circular imports (e.g. File A imports File B, File B imports File C, and File C imports File A). AaC has some basic internal checks to de-conflict circular imports, but circular imports may cause unexpected behavior and indicate poor project organization.
```

In this example, a user unfamiliar with the project can easily mistake `simple_enum` as a primitive type.

## Enumeration Values

> Enumeration values in AaC are all `UPPSERCASE`, following the common style for enumeration values and constants in programming languages such as Python and Java.

Reviewer: ```suggestion
Enumeration values in AaC are all `UPPERCASE`, following the common style for enumeration values and constants in programming languages such as Python and Java.
```

## Root Keys
Like the primitive types, the available root keys in AaC are defined in the definition `Root`. Each field entry in the `Root` definition specifies the root key (the field's name) and a corresponding schema definition (the field type). Additional root keys can be defined by users via an extension to the `Root` definition.

Like primitive types, root keys should follow the `snake_case` style, but emphasize succinct, descriptive names. Example root keys are `import`, `schema`, `model`, etc. If a one-word type isn't descriptive enough for your root key, keeping it as succinct as possible (e.g. `spring_boot_service`, `cloud_gateway`) is preferable. Long-named root keys, or roots keys that aren't `snake_case`, will not conform to the standard employed by the Core Specification and first-party plugins and could introduce confusion about what the root key is.

# Organizing Imports
Each AaC file has a file-wide import declaration definition. Using an `import` definition allows users to reference definitions provided via other user files and can be used to decompose large, complex models into multiple files.

> Because `import` decalarations are file-wide, these clauses should be the first clause in AaC files. There is no internal logic that dictates the order of the entries in an AaC file, but maintaining the `import` clause as the first definition in AaC files makes the file-wide dependencies immediately visible and informs readers on the dependencies of file. Additionally, while it's possible to include multiple `import` definitions in a single file, the preferred approach is to list all imported files into a single `import` definition in each file.

Reviewer: ```suggestion
Because `import` decalarations are file-wide, these clauses should be the first clause in AaC files. There is no internal logic that dictates the order of the entries in an AaC file, but maintaining the `import` clause as the first definition in AaC files makes the file-wide dependencies immediately visible and informs readers on the dependencies of file. Additionally, while it's possible to include multiple `import` definitions in a single file, the preferred approach is to list all imported files in a single `import` definition in each file.
```
